sources/imdsclient/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+/*!
+The imdsclient library provides high-level methods to interact with the AWS Instance Metadata Service.
+The high-level methods provided are [`fetch_dynamic`], [`fetch_metadata`], and [`fetch_userdata`].
+
+For more control, and to query IMDS without high-level wrappers, there is also a [`fetch_imds`] method.
+This method is useful for specifying things like a pinned date for the IMDS schema version.
+*/
+
+#![deny(rust_2018_idioms)]
+
+use http::StatusCode;
+use log::{debug, info, trace, warn};
+use reqwest::blocking::Client;
+use snafu::{ensure, ResultExt};
+
+const IMDS_BASE_URI: &str = "http://169.254.169.254";
+const IMDS_PINNED_DATE: &str = "2020-10-27";
+
+// Currently only able to get fetch session tokens from `latest`
+const IMDS_SESSION_TOKEN_URI: &str = "http://169.254.169.254/latest/api/token";
+
+/// A client for making IMDSv2 queries.
+/// It obtains a session token when it is first instantiated and is reused between helper functions.
+pub struct ImdsClient {
+    client: Client,
+    session_token: String,
+}
+
+impl ImdsClient {
+    pub fn new() -> Result<Self> {
+        let client = Client::new();
+        let session_token = fetch_token(&client)?;
+        Ok(Self {
+            client,
+            session_token,
+        })
+    }
+
+    /// Helper to fetch `dynamic` targets from IMDS, preferring an override file if present.
+    pub fn fetch_dynamic(&mut self, end_target: &str, description: &str) -> Result<Option<String>> {
+        let dynamic_target = format!("dynamic/{}", end_target);
+        self.fetch_imds(IMDS_PINNED_DATE, &dynamic_target, description)
+    }
+
+    /// Helper to fetch `meta-data` targets from IMDS, preferring an override file if present.
+    pub fn fetch_metadata(
+        &mut self,
+        end_target: &str,
+        description: &str,
+    ) -> Result<Option<String>> {
+        let metadata_target = format!("meta-data/{}", end_target);
+        self.fetch_imds(IMDS_PINNED_DATE, &metadata_target, description)
+    }
+
+    /// Helper to fetch `user-data` from IMDS, preferring an override file if present.
+    pub fn fetch_userdata(&mut self) -> Result<Option<String>> {
+        self.fetch_imds(IMDS_PINNED_DATE, "user-data", "user-data")
+    }
+
+    /// Fetch data from IMDS, preferring an override file if present.
+    pub fn fetch_imds(
+        &mut self,
+        schema_version: &str,
+        target: &str,
```

Comment:
Any reason these `&str`s shouldn't be `AsRef<str>`?

Reply:
I didn't think about it, but I think it's a good idea to make that change.